src/drivers/roboclaw/RoboClaw.cpp

```diff
@@ -112,84 +133,189 @@ RoboClaw::~RoboClaw()
 	close(_uart);
 }
 
-int RoboClaw::readEncoder(e_motor motor)
+void RoboClaw::taskMain()
 {
-	uint16_t sum = 0;
-
-	if (motor == MOTOR_1) {
-		_sendCommand(CMD_READ_ENCODER_1, nullptr, 0, sum);
+	// Make sure the Roboclaw is actually connected, so I don't just spam errors if it's not.
+	uint8_t rbuff[4];
+	int err_code = _transaction(CMD_READ_STATUS, nullptr, 0, &rbuff[0], sizeof(rbuff), false, true);
 
-	} else if (motor == MOTOR_2) {
-		_sendCommand(CMD_READ_ENCODER_2, nullptr, 0, sum);
+	if (err_code <= 0) {
+		PX4_ERR("Unable to connect to Roboclaw. Shutting down Roboclaw driver.");
+		return;
 	}
 
-	uint8_t rbuf[50];
-	usleep(5000);
-	int nread = read(_uart, rbuf, 50);
+	// This main loop performs two different tasks, asynchronously:
+	// - Send actuator_controls_0 to the Roboclaw as soon as they are available
+	// - Read the encoder values at a constant rate
+	// To do this, the timeout on the poll() function is used.
+	// waitTime is the amount of time left (int microseconds) until the next time I should read from the encoders.
+	// It is updated at the end of every loop. Sometimes, if the actuator_controls_0 message came in right before
+	// I should have read the encoders, waitTime will be 0. This is fine. When waitTime is 0, poll() will return
+	// immediately with a timeout. (Or possibly with a message, if one happened to be available at that exact moment)
+	uint64_t encoderTaskLastRun = 0;
+	int waitTime = 0;
+
+	_actuatorsSub = orb_subscribe(ORB_ID(actuator_controls_0));
+	orb_set_interval(_actuatorsSub, _parameters.actuator_write_period_ms);
+
+	_armedSub = orb_subscribe(ORB_ID(actuator_armed));
+	_paramSub = orb_subscribe(ORB_ID(parameter_update));
+
+	pollfd fds[3];
+	fds[0].fd = _paramSub;
+	fds[0].events = POLLIN;
+	fds[1].fd = _actuatorsSub;
+	fds[1].events = POLLIN;
+	fds[2].fd = _armedSub;
+	fds[2].events = POLLIN;
```

Comment:
Given you have all these polls, have you checked with `top` what the CPU usage is?

Usually we only poll for one topic, usually the fast one that should trigger everything else and then we copy the other topics if they have changed in the meantime.

Reply:
I changed it to only poll one topic (`actuator_controls_0`) and the CPU usage is unchanged (about 3.4% on a Pixhawk 4), but it causes it to respond slowly to disarming. I can fix this, but I'm not sure it will be worth the effort since it doesn't affect CPU usage at all.

Reply:
Oh interesting. _Never optimize without profiling_ I guess :smile:. Your call!